caffe2/core/operator_c10wrapper.h

```diff
@@ -91,56 +98,44 @@ class C10OperatorWrapper final : public Operator<Context> {
     return Parameter::parse(ArgumentHelper(operator_def));
   }
 
-  template <
-      size_t... InputIndex,
-      size_t... OutputIndex,
-      size_t... ParameterIndex>
-  c10::guts::enable_if_t<
-      details::true_t<InputIndex...>::value &&
-          !use_array_input,
-      void>
-  RunOnDevice_(
-      c10::guts::index_sequence<InputIndex...>,
-      c10::guts::index_sequence<OutputIndex...>,
-      c10::guts::index_sequence<ParameterIndex...>) {
-    Stack stack;
-    torch::jit::push(stack,
-      IValue(at::Tensor(C10Tensor(Input(InputIndex))))...,
-      IValue(std::get<ParameterIndex>(parameters_))...,
-      IValue(at::Tensor(C10Tensor(*Output(OutputIndex))))...
-    );
-    call_(&stack);
-    // TODO Do we have to Write outputs from stack back into the workspace?
+  void pushInputs_() {
+    if (use_array_input) {
+      stack_.emplace_back(ivalue::TensorList::create(array_inputs_()));
+    } else {
+      for (size_t i = 0; i < num_inputs(); ++i) {
+        stack_.emplace_back(at::Tensor(C10Tensor(Input(i))));
+      }
+    }
   }
 
-  template <
-      size_t... InputIndex,
-      size_t... OutputIndex,
-      size_t... ParameterIndex>
-  c10::guts::enable_if_t<
-      details::true_t<InputIndex...>::value &&
-          use_array_input,
-      void>
-  RunOnDevice_(
-      c10::guts::index_sequence<InputIndex...>,
-      c10::guts::index_sequence<OutputIndex...>,
-      c10::guts::index_sequence<ParameterIndex...>) {
-    Stack stack;
-    torch::jit::push(stack,
-      IValue(ivalue::TensorList::create(array_inputs_())),
-      IValue(std::get<ParameterIndex>(parameters_))...,
-      IValue(at::Tensor(C10Tensor(*Output(OutputIndex))))...
-    );
-    call_(&stack);
-    // TODO Do we have to Write outputs from stack back into the workspace?
+  template<size_t... ParameterIndex>
+  void pushParameters_(guts::index_sequence<ParameterIndex...>) {
+    (void)std::initializer_list<int>{(
+      stack_.emplace_back(std::get<ParameterIndex>(parameters_))
+    , 0)...};
+  }
+
+  void pushOutputParameters_() {
+    for (size_t i = 0; i < num_output_parameters; ++i) {
+      stack_.emplace_back(at::Tensor(C10Tensor(*Output(i))));
```

Comment:
are we constructing an uninitialized Tensor here?

Reply:
Yes, unfortunately. I'm currently fixing this.

Reply:
it's fixed in https://github.com/pytorch/pytorch/pull/16643